packages/filecoin-api/test/events/aggregator.js

```diff
@@ -0,0 +1,1236 @@
+    assert.equal(context.service.aggregate?.offer?.callCount, 1)
+    assert.ok(
+      blockPieces.cid.equals(
+        // @ts-expect-error not typed hooks
+        context.service.aggregate?.offer?._params[0].nb.pieces
+      )
+    )
+    assert.ok(
+      aggregateRecord.aggregate.equals(
+        // @ts-expect-error not typed hooks
+        context.service.aggregate?.offer?._params[0].nb.aggregate
+      )
+    )
+  },
+  'handles aggregate insert event errors when fails to read from buffer store':
+    wichMockableContext(
+      async (assert, context) => {
+        const piecesLength = 100
+        const group = context.id.did()
+        const { aggregate, pieces } = await randomAggregate(piecesLength, 128)
+
+        /** @type {Buffer} */
+        const buffer = {
+          pieces: pieces.map((p) => ({
+            piece: p.link,
+            insertedAt: Date.now(),
+            policy: 0,
+          })),
+          group,
+        }
+        const blockBuffer = await CBOR.write(buffer)
+
+        // Put aggregate record
+        const aggregateRecord = {
+          pieces: blockBuffer.cid,
+          aggregate: aggregate.link,
+          group,
+          insertedAt: Date.now(),
+        }
+        const putAggregateRes = await context.aggregateStore.put(
+          aggregateRecord
+        )
+        assert.ok(putAggregateRes.ok)
+
+        // Handle event
+        const handledAggregateInsertsRes =
+          await AggregatorEvents.handleAggregateInsertToAggregateOffer(
+            context,
+            aggregateRecord
+          )
+        assert.ok(handledAggregateInsertsRes.error)
+        assert.equal(
+          handledAggregateInsertsRes.error?.name,
+          StoreOperationErrorName
+        )
+      },
+      async (context) => ({
+        ...context,
+        bufferStore:
+          getStoreImplementations(FailingStore).aggregator.bufferStore,
+      })
+    ),
+}
+
+/**
+ * @param {number} length
+ * @param {string} group
+ * @param {object} [piecesOptions]
+ * @param {number} [piecesOptions.length]
+ * @param {number} [piecesOptions.size]
+ */
+async function getBuffers(length, group, piecesOptions = {}) {
+  const piecesLength = piecesOptions.length || 100
+  const piecesSize = piecesOptions.size || 128
+
+  const pieceBatches = await Promise.all(
+    Array.from({ length }).map(() => randomCargo(piecesLength, piecesSize))
+  )
+
+  const buffers = pieceBatches.map((b) => ({
+    pieces: b.map((p) => ({
+      piece: p.link,
+      insertedAt: Date.now(),
+      policy: /** @type {PiecePolicy} */ (0),
+    })),
+    group,
+  }))
+
+  return {
+    buffers,
+    blocks: await Promise.all(buffers.map((b) => CBOR.write(b))),
+  }
+}
+
+/**
+ * @param {API.Test<AggregatorApi.TestEventsContext>} testFn
+ * @param {(context: AggregatorApi.TestEventsContext) => Promise<AggregatorApi.TestEventsContext>} mockContextFunction
+ */
+function wichMockableContext(testFn, mockContextFunction) {
```

Comment:
A helper?

Reply:
I could not get this to work with the chain of type inference if I put this with templates to try to make it generic with all the contexts being inferred by caller. So keeping it for now to not spend more time